// vue-language-server/dist/utils/prettier/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _ = require("lodash");
const vscode_languageserver_types_1 = require("vscode-languageserver-types");
const strings_1 = require("../strings");
function prettierify(code, filePath, range, initialIndent, formatParams, prettierVSCodeConfig, parser) {
    try {
        const prettier = require('prettier');
        const prettierOptions = getPrettierOptions(prettierVSCodeConfig, parser, filePath);
        const prettierifiedCode = prettier.format(code, prettierOptions);
        return [toReplaceTextedit(prettierifiedCode, range, formatParams, initialIndent)];
    }
    catch (e) {
        console.log('Prettier format failed');
        console.error(e.message);
        return [];
    }
}
exports.prettierify = prettierify;
function prettierEslintify(code, filePath, range, initialIndent, formatParams, prettierVSCodeConfig, parser) {
    try {
        const prettierEslint = require('prettier-eslint');
        const prettierOptions = getPrettierOptions(prettierVSCodeConfig, parser, filePath);
        const prettierifiedCode = prettierEslint({
            filePath: require('path').dirname(filePath),
            text: code,
            fallbackPrettierOptions: prettierOptions
        });
        return [toReplaceTextedit(prettierifiedCode, range, formatParams, initialIndent)];
    }
    catch (e) {
        console.log('Prettier-Eslint format failed');
        console.error(e.message);
        return [];
    }
}
exports.prettierEslintify = prettierEslintify;
function getPrettierOptions(prettierVSCodeConfig, parser, filePath) {
    let trailingComma = prettierVSCodeConfig.trailingComma;
    if (trailingComma === true) {
        trailingComma = 'es5';
    }
    else if (trailingComma === false) {
        trailingComma = 'none';
    }
    const prettierOptions = {
        printWidth: prettierVSCodeConfig.printWidth,
        tabWidth: prettierVSCodeConfig.tabWidth,
        singleQuote: prettierVSCodeConfig.singleQuote,
        trailingComma,
        bracketSpacing: prettierVSCodeConfig.bracketSpacing,
        jsxBracketSameLine: prettierVSCodeConfig.jsxBracketSameLine,
        parser,
        semi: prettierVSCodeConfig.semi,
        useTabs: prettierVSCodeConfig.useTabs,
        arrowParens: prettierVSCodeConfig.arrowParens
    };
    const prettier = require('prettier');
    const prettierrcOptions = prettier.resolveConfig.sync(filePath, { useCache: false });
    if (!prettierrcOptions) {
        return prettierOptions;
    }
    else {
        // The only alternative parser that can be specified is `flow` for <script>
        if (parser === 'babylon') {
            return _.assign(prettierOptions, prettierrcOptions);
        }
        // Otherwise, use the parser specified by Vetur
        return _.assign(prettierOptions, prettierrcOptions, {
            parser
        });
    }
}
function toReplaceTextedit(prettierifiedCode, range, formatParams, initialIndent) {
    if (initialIndent) {
        // Prettier adds newline at the end
        const formattedCode = '\n' + strings_1.indentSection(prettierifiedCode, formatParams);
        return vscode_languageserver_types_1.TextEdit.replace(range, formattedCode);
    }
    else {
        return vscode_languageserver_types_1.TextEdit.replace(range, '\n' + prettierifiedCode);
    }
}
