// vue-language-server/dist/modes/template/tagProviders/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const htmlTags_1 = require("./htmlTags");
const vueTags_1 = require("./vueTags");
const routerTags_1 = require("./routerTags");
const externalTagProviders_1 = require("./externalTagProviders");
var componentTags_1 = require("./componentTags");
exports.getComponentTags = componentTags_1.getComponentTags;
const ts = require("typescript");
const fs = require("fs");
exports.allTagProviders = [
    htmlTags_1.getHTML5TagProvider(),
    vueTags_1.getVueTagProvider(),
    routerTags_1.getRouterTagProvider(),
    externalTagProviders_1.elementTagProvider,
    externalTagProviders_1.onsenTagProvider,
    externalTagProviders_1.bootstrapTagProvider,
    externalTagProviders_1.vuetifyTagProvider
];
function getTagProviderSettings(workspacePath) {
    const settings = {
        html5: true,
        vue: true,
        router: false,
        element: false,
        onsen: false,
        bootstrap: false,
        vuetify: false
    };
    if (!workspacePath) {
        return settings;
    }
    try {
        const packagePath = ts.findConfigFile(workspacePath, ts.sys.fileExists, 'package.json');
        if (!packagePath) {
            return settings;
        }
        const packageJson = JSON.parse(fs.readFileSync(packagePath, 'utf-8'));
        if (packageJson.dependencies['vue-router']) {
            settings['router'] = true;
        }
        if (packageJson.dependencies['element-ui']) {
            settings['element'] = true;
        }
        if (packageJson.dependencies['vue-onsenui']) {
            settings['onsen'] = true;
        }
        if (packageJson.dependencies['bootstrap-vue']) {
            settings['bootstrap'] = true;
        }
        if (packageJson.dependencies['vuetify']) {
            settings['vuetify'] = true;
        }
    }
    catch (e) { }
    return settings;
}
exports.getTagProviderSettings = getTagProviderSettings;
function getEnabledTagProviders(tagProviderSetting) {
    return exports.allTagProviders.filter(p => tagProviderSetting[p.getId()] !== false);
}
exports.getEnabledTagProviders = getEnabledTagProviders;
