// vue-language-server/dist/modes/template/services/htmlFormat.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _ = require("lodash");
const vscode_languageserver_types_1 = require("vscode-languageserver-types");
const js_beautify_1 = require("js-beautify");
const templateHead = '<template>';
const templateTail = '</template>';
function htmlFormat(document, currRange, formattingOptions, config) {
    const { value, range } = getValueAndRange(document, currRange);
    defaultHtmlOptions.indent_with_tabs = !formattingOptions.insertSpaces;
    defaultHtmlOptions.indent_size = formattingOptions.tabSize;
    const htmlFormattingOptions = _.assign(defaultHtmlOptions, config.vetur.format.defaultFormatterOptions['js-beautify-html'], { end_with_newline: false });
    const beautifiedHtml = js_beautify_1.html(templateHead + value + templateTail, htmlFormattingOptions);
    const wrappedHtml = beautifiedHtml.substring(templateHead.length, beautifiedHtml.length - templateTail.length);
    return [
        {
            range,
            newText: wrappedHtml
        }
    ];
}
exports.htmlFormat = htmlFormat;
function getValueAndRange(document, currRange) {
    let value = document.getText();
    let range = currRange;
    if (currRange) {
        const startOffset = document.offsetAt(currRange.start);
        const endOffset = document.offsetAt(currRange.end);
        value = value.substring(startOffset, endOffset);
    }
    else {
        range = vscode_languageserver_types_1.Range.create(vscode_languageserver_types_1.Position.create(0, 0), document.positionAt(value.length));
    }
    return { value, range };
}
const defaultHtmlOptions = {
    brace_style: 'collapse',
    end_with_newline: false,
    indent_char: ' ',
    indent_handlebars: false,
    indent_inner_html: false,
    indent_scripts: 'keep',
    indent_size: 2,
    indent_with_tabs: false,
    max_preserve_newlines: 1,
    preserve_newlines: true,
    unformatted: [
        'area',
        'base',
        'br',
        'col',
        'embed',
        'hr',
        'img',
        'input',
        'keygen',
        'link',
        'menuitem',
        'meta',
        'param',
        'source',
        'track',
        'wbr'
    ],
    wrap_line_length: 0,
    wrap_attributes: 'auto'
    // Wrap attributes to new lines [auto|force|force-aligned|force-expand-multiline] ["auto"]
};
